.github/workflows/main.yml

```diff
@@ -0,0 +1,62 @@
+name: CI
+
+# Controls when the action will run. Triggers the workflow on push or pull request
+# events but only for the master branch
+on:
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  build:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-16.04, ubuntu-18.04, ubuntu-20.04]
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        # we don't need submodules for building the toolchain, but we do for everything else
+        with:
+          submodules: 'recursive'
+          fetch-depth: 0 # submodule checkout does not work with depth=1 (the default)
+
+      # needed both at compile- and runtime for the toolchain
+      - name: Install toolchain dependencies
+        run: sudo apt install -y libgmp-dev libmpfr-dev libmpc-dev
+
+      # build the toolchain first
+      # it takes ~40min to build, so we want to avoid that whenever possible!
+      - uses: actions/cache@v2
+        id: toolchain-cache
+        with:
+          path: |
+            toolchain.tar.gz
```

Comment:
Does this save the toolchain twice now?

Reply:
Yes, once for further use, and once for the artifact uploaded later. I suppose we could just offer the install directory as an artifact instead of making a .tar.gz

Reply:
Actually, we should see if the day-to-day runtime changes when we're uploading a directory as an artifact, instead of a cached .tar.gz.
A job is running right now, but it's rebuilding the toolchain (because the paths changed): https://github.com/shinyblink/Ndless/actions/runs/196901972
I will trigger another run later, to get a runtime with the new artifact upload.

If the runtime regresses meaningfully, I would revert to how it was before - the tar.gz doesn't hurt anything, and only gets made when the toolchain is built anyway.

Reply:
The results are in:

With uploading a cached `toolchain.tar.gz` (~400-450MB):
https://github.com/shinyblink/Ndless/actions/runs/196901382

=> about 3:45 minutes

With uploading the cached toolchain install directory as-is (~1-1.5 GB):
https://github.com/shinyblink/Ndless/actions/runs/196901972

=> about 4:45 minutes

The previous (duplicating) approach wins by a minute, and with much smaller artifacts.

I'll revert to the tar.gz method.

Reply:
What's stored in the cache is separate from artifacts.

I intended the tar.gz upload to be kind of a "nightly" toolchain build, but I guess releases are the better tool for the job. Github Actions don't let you reuse past build's artifacts, because they're not really intended to be used like that. That's why the packaged tar.gz toolchain is also cached next to the unextracted toolchain, so it just has to be uploaded again.

I think I will remove the toolchain artifact upload though, because it does take a long time. Maybe I'll come back around to implement nightly builds in some way, but I'm not quite sure how that will work with Actions.